extensions/wikia/Recirculation/js/incontent.js

```diff
@@ -0,0 +1,187 @@
+
+
+	var $container = $('#mw-content-text'),
+		template;
+
+	function injectInContentWidget($container) {
+		var $links = $container.find('a'),
+			sections = buildSections($container.find('h2')),
+			width,
+			firstSuitableSection,
+			topTitles;
+
+		// If this page doesn't have enough content (either links or sections) we
+		// don't want to show this widget
+		if ($links.length < 8 || sections.length < 3) {
+			return;
+		}
+
+		// The idea is to show links above the first section under an infobox
+		width = $container.outerWidth();
+		firstSuitableSection = sections.find(function(item, index) {
+			return item.width === width;
+		});
+
+		if (!firstSuitableSection) {
+			return;
+		}
+
+		topTitles = findTopTitles($links);
+
+		if (topTitles.length < 3) {
+			return;
+		}
+
+		nirvana.sendRequest({
+			controller: 'ArticlesApi',
+			method: 'getDetails',
+			format: 'json',
+			type: 'get',
+			data: {
+				titles: topTitles.join(','),
+				abstract: 0,
+				width: 50,
+				height: 50
+			},
+			callback: renderArticles(firstSuitableSection)
+		});
+	}
+
+	function renderArticles(section) {
+		var placeholderImage = w.wgExtensionsPath + '/wikia/Recirculation/images/placeholder.png';
+		tracker.trackImpression('in-content');
+
+		return function (response) {
+			var items = [],
+				html;
+
+			$.each(response.items, function(index, item) {
+				item.thumbnail = item.thumbnail || placeholderImage;
+				items.push(item);
+			});
+
+			$html = $(Mustache.render(template, {
+				title: $.msg('recirculation-incontent-title'),
+				items: items
+			}));
+
+			section.$start.before($html);
+
+			$html.on('mousedown', 'a', function() {
+				tracker.trackClick('in-content');
+			});
+		}
+	}
+
+	/**
+	 * Build the sections array
+	 *
+	 * DOM/layout querying: OK
+	 * DOM modification:    forbidden
+	 *
+	 * @param {jQuery} $headers headers dividing the article to sections
+	 * @returns {Section[]}
+	 */
+	function buildSections($headers) {
+		var i,
+			len,
+			sections = [],
+			intro,
+			$start,
+			$end,
+			section;
+
+		for (i = 0, len = $headers.length; i < len + 1; i += 1) {
+			intro = (i === 0);
+			$start = !intro && $headers.eq(i - 1);
+			$end = $headers.eq(i);
+			section = {
+				intro: intro,
+				$start: intro ? undefined : $start,
+				$end: $end
+			};
+			section.width = $start && $start.outerWidth();
+
+			sections.push(section);
+		}
+
+		return sections;
+	}
+
+	function findTopTitles($links) {
+		var links = buildLinks($links),
+			titles = getSortedKeys(links),
+			topTitles = [],
+			i = 0;
+
+		while (topTitles.length < 3) {
+			if (titles[i]) {
+				topTitles.push(titles[i]);
+			}
+			i ++;
+		}
+
+		return topTitles;
+	}
+
+	function validLink(element) {
+		// Not a link to current article
+		if (element.title === w.wgTitle) {
+			return false;
+		}
+
+		// This is a pretty heavy handed Regex, and it may only work for EN communities
+		// but the idea is to not display links to special pages
```

Comment:
Do you mean special pages as the one in `Special` namespace (for example: `Special:WikiActivity`) or all pages within other than regular (`NS_MAIN`) namespace?